File: solutions/2026-03-08/cpp/2026-03-08-find-unique-binary-string.cpp
// Problem: Find Unique Binary String
// Summary: Given a list of unique binary strings of length n, find a binary string of length n not present in the list.
// Link: https://leetcode.com/problems/find-unique-binary-string/
// Approach:
// We can use the property that there are 2^n possible binary strings of length n.
// Since the input array `nums` contains `n` unique binary strings, and `n` is at most 16,
// the total number of possible binary strings (2^n) is at most 2^16 = 65536, which is manageable.
// We can convert each binary string in `nums` to its integer representation.
// Then, we iterate through all possible integers from 0 to 2^n - 1.
// For each integer, we convert it back to its binary string representation of length `n`.
// If this generated binary string is not found in the set of existing binary strings (from `nums`),
// we return it as the unique binary string.
//
// To efficiently check for the presence of a binary string, we can store all binary strings from `nums`
// in a hash set (std::unordered_set<std::string> in C++).
//
// Alternatively, a more direct and elegant approach leverages Cantor's diagonalization argument.
// We can construct a unique binary string by taking the i-th character of the i-th string in `nums`
// and flipping it (0 becomes 1, 1 becomes 0).
// For example, if `nums = ["01", "10"]`:
// - For the 0-th string "01", we take the 0-th character '0' and flip it to '1'.
// - For the 1-st string "10", we take the 1-st character '0' and flip it to '1'.
// The resulting string is "11". This string is guaranteed to be different from all strings in `nums`
// because its i-th character differs from the i-th character of the i-th string in `nums`.
//
// Time Complexity:
// Using the diagonalization approach: O(n^2) to iterate through the strings and their characters.
// Using the set approach: O(n^2) to insert all strings into the set and O(n^2) in the worst case
// to generate and check for each potential unique string. Overall, O(n^2).
//
// Space Complexity:
// Using the diagonalization approach: O(n) to store the result string.
// Using the set approach: O(n^2) to store all strings in the hash set.
//
// We will implement the diagonalization approach for its efficiency and simplicity.

#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>

class Solution {
public:
    std::string findDifferentBinaryString(std::vector<std::string>& nums) {
        // n represents the length of each binary string and the number of strings in nums.
        int n = nums.size();
        // The result string will store the unique binary string we construct.
        std::string result = "";

        // Iterate through each position (index i) from 0 to n-1.
        for (int i = 0; i < n; ++i) {
            // For the i-th string in `nums` (which is `nums[i]`),
            // we consider its i-th character (`nums[i][i]`).
            // We flip this character to construct the i-th character of our result string.
            // If the character is '0', we append '1' to the result.
            // If the character is '1', we append '0' to the result.
            if (nums[i][i] == '0') {
                result += '1';
            } else {
                result += '0';
            }
        }

        // The constructed `result` string is guaranteed to be unique.
        // This is because for any string `nums[i]` in the input, the i-th character of `result`
        // is different from the i-th character of `nums[i]`.
        // Therefore, `result` cannot be equal to any string in `nums`.
        return result;
    }
};

File: solutions/2026-03-08/python/2026-03-08-find-unique-binary-string.py
```python
# Problem: Find Unique Binary String
# Summary: Given n unique binary strings of length n, find one binary string of length n that is not present.
# Link: https://leetcode.com/problems/find-unique-binary-string/
#
# Approach:
# The problem guarantees that there are n unique binary strings of length n in the input array `nums`.
# There are a total of 2^n possible binary strings of length n.
# Since n is at most 16, 2^n is at most 2^16 = 65536, which is a manageable number.
# We can iterate through all possible binary strings of length n and check if each string exists in `nums`.
# The first string we find that is NOT in `nums` is our answer.
#
# A simpler and more efficient approach leverages Cantor's diagonalization argument, similar to how one might prove that the set of real numbers is uncountable.
# We can construct a unique binary string by taking the i-th character of the i-th string in `nums` and inverting it.
# For example, if nums = ["01", "10"]:
# - For the first string "01", take its 0-th character: '0'. Invert it to '1'.
# - For the second string "10", take its 1-st character: '0'. Invert it to '1'.
# This gives us "11".
# This constructed string is guaranteed to be unique because it differs from every string in `nums` at at least one position (specifically, the position corresponding to the string's index).
#
# Time Complexity:
# O(n^2) - We iterate through `nums` once (n strings). For each string, we access its i-th character. String indexing is O(1). Concatenating characters to form the result string takes O(n). So, constructing the unique string is O(n).
# If we were to use the brute-force approach of checking all 2^n strings, the complexity would be much higher.
#
# Space Complexity:
# O(n) - To store the resulting unique binary string.

class Solution:
    def findDifferentBinaryString(self, nums: list[str]) -> str:
        n = len(nums)
        unique_string = []

        # Iterate through each string in nums by index
        for i in range(n):
            # Take the i-th character of the i-th string
            char_to_invert = nums[i][i]
            
            # Invert the character ('0' becomes '1', '1' becomes '0')
            inverted_char = '1' if char_to_invert == '0' else '0'
            
            # Append the inverted character to our unique string builder
            unique_string.append(inverted_char)
        
        # Join the list of characters into a single string
        return "".join(unique_string)

```